src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs

```diff
@@ -41,37 +58,155 @@ public partial class Ping
             }
         }
 
-        private async Task<PingReply> SendIcmpEchoRequestOverRawSocket(IPAddress address, byte[] buffer, int timeout, PingOptions options)
+        private SocketConfig GetSocketConfig(IPAddress address, byte[] buffer, int timeout, PingOptions options)
         {
-            EndPoint endPoint = new IPEndPoint(address, 0);
+            SocketConfig config = new SocketConfig();
+            config.EndPoint = new IPEndPoint(address, 0);
+            config.Timeout = timeout;
+            config.Options = options;
 
-            bool isIpv4 = address.AddressFamily == AddressFamily.InterNetwork;
-            ProtocolType protocolType = isIpv4 ? ProtocolType.Icmp : ProtocolType.IcmpV6;
+            config.IsIpv4 = address.AddressFamily == AddressFamily.InterNetwork;
+            config.ProtocolType = config.IsIpv4 ? ProtocolType.Icmp : ProtocolType.IcmpV6;
 
-            // Use a random value as the identifier.  This doesn't need to be perfectly random
+            // Use a random value as the identifier. This doesn't need to be perfectly random
             // or very unpredictable, rather just good enough to avoid unexpected conflicts.
             Random rand = t_idGenerator ?? (t_idGenerator = new Random());
-            ushort identifier = (ushort)rand.Next((int)ushort.MaxValue + 1);
+            config.Identifier = (ushort)rand.Next((int)ushort.MaxValue + 1);
 
             IcmpHeader header = new IcmpHeader()
             {
-                Type = isIpv4 ? (byte)IcmpV4MessageType.EchoRequest : (byte)IcmpV6MessageType.EchoRequest,
+                Type = config.IsIpv4 ? (byte)IcmpV4MessageType.EchoRequest : (byte)IcmpV6MessageType.EchoRequest,
                 Code = 0,
                 HeaderChecksum = 0,
-                Identifier = identifier,
+                Identifier = config.Identifier,
                 SequenceNumber = 0,
             };
 
-            byte[] sendBuffer = CreateSendMessageBuffer(header, buffer);
+            config.SendBuffer = CreateSendMessageBuffer(header, buffer);
+            return config;
+        }
+
+        private Socket GetRawSocket(SocketConfig socketConfig)
+        {
+            // Setting Socket.DontFragment and .Ttl is not supported on Unix, so socketConfig.Options is ignored.
```

Comment:
@wfurt, @tdinucci, seems like your changes (with https://github.com/dotnet/corefx/pull/30324) may conflict.

Reply:
I can wait. I decided to do more work on unit test.